README.md
[![Build](https://travis-ci.org/e-sites/ESOpenSourceLicensesKit.svg)](https://travis-ci.org/e-sites/ESOpenSourceLicensesKit)
[![Platform](https://cocoapod-badges.herokuapp.com/p/ESOpenSourceLicensesKit/badge.png)](http://cocoadocs.org/docsets/ESOpenSourceLicensesKit)
[![Version](https://cocoapod-badges.herokuapp.com/v/ESOpenSourceLicensesKit/badge.png)](http://cocoadocs.org/docsets/ESOpenSourceLicensesKit)
[![Quality](https://apps.e-sites.nl/cocoapodsquality/ESOpenSourceLicensesKit/badge.svg?002)](https://cocoapods.org/pods/ESOpenSourceLicensesKit/quality)

# ESOpenSourceLicensesKit

*Requires Xcode 7.0 and Swift 2.0*

A script to automatically generate an HTML file of all the license files used with your CocoaPods Project

## Example

See [example-output.html](http://htmlpreview.github.io/?https://github.com/e-sites/ESOpenSourceLicensesKit/blob/master/Examples/example-output.html) for an example output

![](Assets/pod-update.gif) ![](Assets/example.gif) 

## Usage

### Cocoapods

Add the following lines to your PodFile:

```ruby
pod 'ESOpenSourceLicensesKit'

post_install do |installer_representation|
    system("ruby Pods/ESOpenSourceLicensesKit/ESOpenSourceLicensesKit/Scripts/generate_licenses.rb --post_install")
end
```
Run `pod update` or `pod install`

#### Swift
And then open a `ESOpenSourceLicensesViewController`:

```swift
import ESOpenSourceLicensesKit
        
let vc = ESOpenSourceLicensesViewController()
let navvc = UINavigationController(rootViewController: vc)
self.presentViewController(navvc, animated: true) { _ in
    
}
```

#### Objective-c (not recommended)

Make a swift / objc bridge.

And then open a `ESOpenSourceLicensesViewController`:

```objective-c
#import <ESOpenSourceLicensesViewController.h>

ESOpenSourceLicensesViewController *vc = [ESOpenSourceLicensesViewController new];
UINavigationController *navVc = [[UINavigationController alloc] initWithRootViewController:vc];
[self presentViewController:navVc animated:YES completion:nil];
```


### Manual 
Remember that the project depends on CocoaPods. Because it searches for LICENSE files within the `Pods` directory.
But if you prefer not to use this pod, you can manually call the bash script from your terminal:

```bash
cd /Path/to/xcode-workspace/
ruby ./generate_licenses.rb --input licenses.html
```

Then drag/drop the licenses.html file to your project and you can use it to load in a `UIWebView`

## Customization

```swift
let openSourceView = ESOpenSourceLicensesView()
openSourceView.licenseBackgroundColor = UIColor.blackColor()
openSourceView.licenseTextColor = UIColor.whiteColor()
```

- See [e-sites.github.io/ESOpenSourceLicensesKit](http://e-sites.github.io/ESOpenSourceLicensesKit/Classes.html)


## Todo

- [x] CocoaPods
- [x] Customization of the `ESOpenSourceLicensesKit`
- [x] Swift 1.2
- [x] Swift 2.0
- [ ] A .plist file of all the licenses
- [ ] An array representation of the license files


CHANGELOG.md
# Change Log
All notable changes to this project will be documented in this file. `ESOpenSourceLicensesKit` adheres to [Semantic Versioning](http://semver.org/).

---

## [1.5](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.5) / 2015-09-15
- Swift 2.0

## [1.4](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.4) / 2015-08-25
- Changed from bash script to ruby script

## [1.3.3](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.3.2) / 2015-08-21
- Renamed to `ESOpenSourceLicensesKit`

## [1.3.2](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.3.2) / 2015-08-21
- (Again) improved inline documentation

## [1.3.1](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.3.1) / 2015-08-21
- Improved inline documentation
- Added Objc support

## [1.3](https://github.com/e-sites/ESOpenSourceLicensesKit/releases/tag/1.3) / 2015-08-21
- Transferred to (swift-only) `ESOpenSourceKit`

## 1.2 / 2015-08-19
- Give the view a padding

## 1.1.2 / 2015-08-19
- Swift resources fallback issue fixed
- Allow ESOpenSourceLicenses to be included in the licenses HTML

## 1.1.1 / 2015-08-19
- Subspecs fixed for Swift projects

## 1.1 / 2015-08-19
- Added font styling
- Added text-color styling
- Added background-color styling
- Added border styling
- Improved unit tests
- Swift 2.0

## 1.0.1 / 2015-08-18
- Converted project to ARC

## 1.0.0 / 2015-08-18
- Initial release

ESOpenSourceLicensesKit/Classes/ESOpenSourceLicensesView.swift
//
//  ESOpenSourceLicensesView.swift
//  ESOpenSourceLicensesKit
//
//  Created by Bas van Kuijck on 19-08-15.
//  Copyright © 2015 e-sites. All rights reserved.
//

import Foundation
import UIKit

/**
A UIWebView presentation of the licenses html file.
Which is generated by the bashscript
*/
public class ESOpenSourceLicensesView : UIWebView {
    
    // MARK: - Constructor
    // ____________________________________________________________________________________________________________________
    
    /**
    A convenience initializer
    This will invoke init(frame:)
    
    - returns a ESOpenSourceLicensesView instance
    */
    public convenience init() {
        self.init(frame: CGRectZero)
    }
    
    override init(frame: CGRect) {
        super.init(frame: frame)
        _init()
    }
    
    /**
    Returns an object initialized from data in a given unarchiver.
    
    - parameter: aDecoder	An unarchiver object.
    
    - returns:  `self`, initialized using the data in decoder.
    */
    public required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)
        _init()
    }
    
    private func _init() {
        self.backgroundColor = UIColor.whiteColor()
        self.dataDetectorTypes = .None
        reload()
    }
    
    
    // MARK: - Style
    // ____________________________________________________________________________________________________________________
    
    /**
    The font to be used for the headers
    Default: Helvetica Neue; 16pt
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 18/08/2015
    */
    public var headerFont = UIFont(name: "HelveticaNeue", size: 16)!
    
    /**
    The font to be used for the license text
    Default: Menlo; 12pt
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 18/08/2015
    */
    public var licenseFont = UIFont(name: "Menlo-Regular", size: 12)!
    
    /**
    The text color of the headers
    Default: Black
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    public var headerTextColor:UIColor = UIColor.blackColor()
    
    /**
    The text color of the license text
    Default: Black
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    public var licenseTextColor = UIColor.blackColor()
    
    /**
    The backgroundcolor of the license text
    Default: #EEE
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    public var licenseBackgroundColor = UIColor(white: 0.9333, alpha: 1.0)
    
    /**
    The border color of the license text
    Default: #DDD
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    public var licenseBorderColor = UIColor(white: 0.8666, alpha: 1.0)
    
    /**
    The width of the border of the license text
    Default 1.0
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date 19/08/2015
    */
    public var licenseBorderWidth:CGFloat = 1.0
    
    /**
    The padding in the view
    Default 10.0
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.2
    - date: 19/08/2015
    */
    public var padding:CGFloat = 10.0
    
    // MARK: - Reloading
    // ____________________________________________________________________________________________________________________
    
    /**
    Reloads the license text
    This should be done after a style attribute is changed
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    override public func reload() {
        do {
            var bundle:NSBundle? = nil
            #if TESTS
                bundle = NSBundle(forClass: self.dynamicType)
                #else
                // Try to find ESOpenSourceLicensesKit.bundle
                let ar = [ "Frameworks", "ESOpenSourceLicensesKit.framework", "ESOpenSourceLicensesKit" ]
                var bundlePath:String? = nil
                for i in 0...2 {
                    let nar = ar[i...2]
                    bundlePath = NSBundle.mainBundle().pathForResource(nar.joinWithSeparator("/"), ofType: "bundle")
                    if (bundlePath != nil) {
                        break
                    }
                }
                if (bundlePath == nil) {
                    return
                }
                bundle = NSBundle(path: bundlePath!)
            #endif
            
            
            let path = bundle!.pathForResource("opensource-licenses", ofType: "html")!
            let contents = try NSString(contentsOfFile: path, encoding: NSUTF8StringEncoding)
            let regex = try NSRegularExpression(pattern: "<style>.+?</style>", options: .CaseInsensitive)
            
            let bgRGB = _rgbaFromUIColor(self.backgroundColor!)
            let blockRGB = _rgbaFromUIColor(self.licenseBackgroundColor)
            let borderRGB = _rgbaFromUIColor(self.licenseBorderColor)
            let headerTextRGB = _rgbaFromUIColor(self.headerTextColor)
            let licenseTextRGB = _rgbaFromUIColor(self.licenseTextColor)
            
            let template = NSString(format: "<style> body { background-color: %@; margin:%.0fpx; } p { font-family:'%@'; margin-bottom:10px; display:block; background-color:%@; border:%.0fpx solid %@; font-size:%.0fpx; padding:5px; color:%@; } h2 { font-family: '%@'; font-size:%.0fpx; color:%@; } </style>",
                bgRGB, self.padding,
                self.licenseFont.fontName, blockRGB, self.licenseBorderWidth, borderRGB, self.licenseFont.pointSize, licenseTextRGB,
                self.headerFont.fontName, self.headerFont.pointSize, headerTextRGB)
            
            let modifiedString = regex.stringByReplacingMatchesInString(contents as String, options: .WithoutAnchoringBounds, range: NSMakeRange(0, contents.length), withTemplate: template as String)
            self.loadHTMLString(modifiedString, baseURL: nil)
            
        } catch { }
    }
    
    // MARK: - Helpers
    // ____________________________________________________________________________________________________________________
    
    /**
    Helper function to convert a UIColor to an @"rgba()" string
    
    - parameter:  color	UIColor
   
    - returns: HTML String representation for the color > `rgba(#,#,#,#)`
    
    - author: Bas van Kuijck <bas@e-sites.nl>
    - since: 1.1
    - date: 19/08/2015
    */
    
    private func _rgbaFromUIColor(color: UIColor) -> NSString {
        var red:CGFloat = 0
        var green:CGFloat = 0
        var blue:CGFloat = 0
        var alpha:CGFloat = 0
        color.getRed(&red, green: &green, blue: &blue, alpha: &alpha)
        return NSString(format: "rgba(%.0f, %.0f, %.0f, %.0f)", red * 255, green * 255, blue * 255, alpha)
    }
}